test/core/client_channel/lb_policy/pick_first_test.cc

```diff
@@ -254,45 +255,45 @@ TEST_F(PickFirstTest, GoesIdleWhenConnectionFailsThenCanReconnect) {
 TEST_F(PickFirstTest, WithShuffle) {
   testing::ScopedExperimentalEnvVar env_var(
       "GRPC_EXPERIMENTAL_PICKFIRST_LB_CONFIG");
-  // 6 addresses have 6! = 720 permutations or 0.1% chance that the shuffle
-  // returns initial sequence
   constexpr std::array<absl::string_view, 6> kAddresses = {
       "ipv4:127.0.0.1:443", "ipv4:127.0.0.1:444", "ipv4:127.0.0.1:445",
       "ipv4:127.0.0.1:446", "ipv4:127.0.0.1:447", "ipv4:127.0.0.1:448"};
-  absl::Status status = ApplyUpdate(
-      BuildUpdate(kAddresses, MakePickFirstConfig(true)), lb_policy_.get());
-  EXPECT_TRUE(status.ok()) << status;
-  std::vector<absl::string_view> prev_attempt_connect_order;
-  GetOrderAddressesArePicked(kAddresses, &prev_attempt_connect_order);
-  // There is 0.1% chance this check fails by design. Not an assert to prevent
-  // flake
-  if (absl::MakeConstSpan(prev_attempt_connect_order) ==
-      absl::MakeConstSpan(kAddresses)) {
-    gpr_log(GPR_INFO, "Address order did not change");
-  }
-  constexpr size_t kMaxAttempts = 5;
-  bool shuffled = false;
-  for (size_t attempt = 0; attempt < kMaxAttempts; ++attempt) {
-    std::vector<absl::string_view> address_order;
-    GetOrderAddressesArePicked(kAddresses, &address_order);
-    if (address_order != prev_attempt_connect_order) {
-      shuffled = true;
+  // 6 addresses have 6! = 720 permutations or roughly 0.14% chance that
+  // the shuffle returns same permutation. We allow for several tries to
+  // prevent flake test.
+  constexpr size_t kMaxTries = 10;
+  std::vector<absl::string_view> addresses_after_update;
+  for (size_t i = 0; i < kMaxTries; i++) {
+    absl::Status status = ApplyUpdate(
+        BuildUpdate(kAddresses, MakePickFirstConfig(true)), lb_policy_.get());
+    EXPECT_TRUE(status.ok()) << status;
+    GetOrderAddressesArePicked(kAddresses, &addresses_after_update);
+    if (absl::MakeConstSpan(addresses_after_update) !=
+        absl::MakeConstSpan(kAddresses)) {
       break;
     }
-    prev_attempt_connect_order = std::move(address_order);
   }
-  ASSERT_TRUE(shuffled) << "Addresses are not reshuffled";
+  // Same elements, different order
+  ASSERT_THAT(addresses_after_update,
```

Comment:
These two checks seem unnecessary.  Instead, why not just set a bool on line 273 before we break out of the loop, and then check the bool here?

Reply:
I think that `GetOrderAddressesArePicked()` already guarantees that all addresses will be present, because it removes the subchannel for each address from its map as it sees that subchannel connect, and it keeps going until the map is empty.  So I think this check is redundant.

Reply:
There's now only one check here... I would like to have an explicit check, mostly to protect from an error when writing the test.

Reply:
I don't see how `GetOrderAddressesArePicked()` could produce any other result and still do its job, so I'm not sure what kind of error we're protecting against here.  And if we are worried about that, why check this only after the last iteration of the loop?  Wouldn't you want to check this inside the loop, so that we know that `GetOrderAddressesArePicked()` returns the full set every time?

This seems odd and not useful to me, but I'll leave it up to you.

Reply:
Removed.